codelabs/Utopia Wholesale Codelab/Advanced Features/Deploying to Google Cloud Run.md

```diff
@@ -0,0 +1,279 @@
+
+### **Step 2: Build the Fat JAR**
+
+Build the fat JAR file that includes all dependencies:
+
+```bash
+./gradlew :multipaz-openid4vci-server:buildFatJar
+```
+
+**What this does:**
+- Compiles the Kotlin code
+- Packages all dependencies into a single JAR file
+- Includes the configuration files (including `default_configuration.json`)
+- Outputs the JAR to `multipaz-openid4vci-server/build/libs/app.jar`
+
+**Note:** The Ktor Gradle plugin creates a fat JAR named `app.jar` by default.
+
+### **Step 3: Prepare for Cloud Build**
+
+Copy the JAR file to the deployment directory:
+
+```bash
+# Rename and copy the JAR file to the multipaz-openid4vci-server directory
+cp multipaz-openid4vci-server/build/libs/app.jar multipaz-openid4vci-server/app.jar
+```
+
+**Directory structure should look like:**
+```
+multipaz-openid4vci-server/
+├── Dockerfile
+└── app.jar
+```
+
+### **Step 4: Build Docker Image**
+
+Build the Docker image using Google Cloud Build:
+
+```bash
+gcloud builds submit multipaz-openid4vci-server --tag gcr.io/engaged-list-481123-n3/multipaz-openid4vci-server
+```
+
+**Explanation:**
+- `gcloud builds submit`: Submits a build to Google Cloud Build
+- `multipaz-openid4vci-server`: The directory containing the Dockerfile and JAR file
+- `--tag`: Tags the resulting image with the specified name
+- `gcr.io/engaged-list-481123-n3/`: Your GCP project's Container Registry path
+  - Replace `engaged-list-481123-n3` with your actual GCP project ID
+
+**What happens:**
+1. Cloud Build reads the Dockerfile
+2. Builds a Docker image containing the JAR file
+3. Pushes the image to Google Container Registry
+4. The image is ready to be deployed to Cloud Run
+
+### **Step 5: Deploy to Cloud Run**
+
+Deploy the containerized application to Cloud Run:
+
+```bash
+gcloud run deploy multipaz-openid4vci-server \
+  --image gcr.io/engaged-list-481123-n3/multipaz-openid4vci-server \
+  --platform managed \
+  --region us-central1 \
+  --allow-unauthenticated
+```
+
+**Parameters explained:**
+- `--image`: Specifies the Docker image to deploy (from Step 4)
+- `--platform managed`: Uses Google's fully managed Cloud Run platform
+- `--region us-central1`: Deploys to the specified region
+  - Choose a region close to your users for better performance
+- `--allow-unauthenticated`: Makes the service publicly accessible
+  - Remove this flag if you want to require authentication
+
+**After deployment:**
+- Cloud Run will provide a URL like: `https://multipaz-openid4vci-server-XXXXX.us-central1.run.app`
+- Update the `base_url` in your configuration file with this URL
+- Rebuild and redeploy if you need to update the base URL
+
+---
+
+## **Deploying the Records Server**
+
+Follow the same steps for `multipaz-records-server`:
+
+1. Create a similar Dockerfile in `multipaz-records-server/`
+2. Build the fat JAR: `./gradlew :multipaz-records-server:buildFatJar`
+3. Copy the JAR: `cp multipaz-records-server/build/libs/app.jar multipaz-records-server/app.jar`
```

Comment:
won't this be `cp app.jar multipaz-records-server/app.jar` since the user uploaded it?